### PR TITLE
feat: (levm) Implement different call types

### LINKED ISSUE #530: Opcode: CALLCODE

### LINKED ISSUE #532: Opcode: DELEGATECALL

### LINKED ISSUE #534: Opcode: STATICCALL

### PR DESCRIPTION
**Pull request by iinaki**
**Motivation**

Implement `CALLCODE`, `DELEGATECALL` and `STATICCALL`

**Description**
- `CALL`, `STATICCALL`, `DELEGATECALL` and `CALLCODE`  are similar but have some key differences, for example:
1) `StaticCall` does the same as `Call`, but changes the config of the `CallFrame` to be static
2) `DelegateCall` is used to execute code from another contract, but changing the storage of the actual address. This is done by having an `Option<Address>` in the `CallFrame`. `CallCode` does the same as `DelegateCall` but changes value and msg_sender.
- Created the `generic_call` function to provide a standard on performing calls.

more info in [EVM Deep Dives](https://noxx.substack.com/p/evm-deep-dives-the-path-to-shadowy-a5f?utm_source=%2Fprofile%2F80455042-noxx&utm_medium=reader2&s=r)

Closes #530 
Closes #532 
Closes #534

### CODE REVIEW

File: crates/levm/tests/tests.rs [hunk 1]
```diff
@@ -1162,6 +1161,455 @@ fn mload_uninitialized_memory() {
...
+    ];
+
+    let mut vm = new_vm_with_ops_addr_bal(
+        &caller_ops,
+        Address::from_low_u64_be(U256::from(1).low_u64()),
+        U256::zero(),
+    );
+
+    vm.add_account(callee_address, callee_account);
+    println!("to excec");
```
Comment (belfortep) [inline crates/levm/tests/tests.rs:1191]:
```suggestion
```
I think we should delete the debug prints in tests

File: crates/levm/tests/tests.rs [hunk 2]
```diff
@@ -1162,6 +1161,455 @@ fn mload_uninitialized_memory() {
...
+        Address::from_low_u64_be(U256::from(1).low_u64()),
+        U256::zero(),
+    );
+
+    vm.add_account(callee_address, callee_account);
+
+    vm.execute();
+
+    let current_call_frame = vm.current_call_frame_mut();
+    println!("{:?}", current_call_frame);
```
Comment (belfortep) [inline crates/levm/tests/tests.rs:1229]:
```suggestion
```

File: crates/levm/tests/tests.rs [hunk 3]
```diff
@@ -1162,6 +1161,455 @@ fn mload_uninitialized_memory() {
...
+
+    assert_eq!(U256::from_big_endian(&return_data), U256::from(0xAAAAAAA));
+    assert!(current_call_frame.is_static);
+}
+
+#[test]
+fn delegatecall_changes_own_storage_and_regular_call_doesnt() {
+    // --- DELEGATECALL --- changes account 1 storage
+    let callee_return_value = U256::from(0xBBBBBBB);
+    // let callee_bytecode = callee_return_bytecode(callee_return_value);
```
Comment (belfortep) [inline crates/levm/tests/tests.rs:1391]:
```suggestion
```

**Review by belfortep — APPROVED** [review]
LGTM! Very nice work, just marked some little details

File: crates/levm/tests/tests.rs (hunk 1 above)
Comment (iinaki) [reply crates/levm/tests/tests.rs]: done!

File: crates/levm/tests/tests.rs (hunk 2 above)
Comment (iinaki) [reply crates/levm/tests/tests.rs]: done!

File: crates/levm/tests/tests.rs (hunk 3 above)
Comment (iinaki) [reply crates/levm/tests/tests.rs]: done!

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**